File: ДЗ2 Хан Роман Олегович БПИ207/Program/source/shape.cpp
// shape.cpp - исходный файл, содержащий в себе реализацию методов класса Shape,
// который находится в файле shape.h.
// Данные методы содержат логику обработки обобщенной фигуры и создания специфических.

#include "../headers/triangle.h"
#include "../headers/circle.h"
#include "../headers/rectangle.h"

/**
 * Было решено отнести метод написания цвета в файл также к абстрактному классу.
 * Это помогает не копировать один и тот же функционал в различные различных альтернатив.
 * @param outFile
 */
void Shape::WriteColorInFile(FILE *outFile) {
    switch (color) {
        case RED:
            fprintf(outFile, "\tColor: RED\n");
            break;
        case ORANGE:
            fprintf(outFile, "\tColor: ORANGE\n");
            break;
        case YELLOW:
            fprintf(outFile, "\tColor: YELLOW\n");
            break;
        case GREEN:
            fprintf(outFile, "\tColor: GREEN\n");
            break;
        case LIGHTBLUE:
            fprintf(outFile, "\tColor: LIGHTBLUE\n");
            break;
        case BLUE:
            fprintf(outFile, "\tColor: BLUE\n");
            break;
        case PURPLE:
            fprintf(outFile, "\tColor: PURPLE\n");
            break;
        default:
            fprintf(outFile, "\tColor: TRANSPARENT\n");
            break;
    }
}

/**
 * Геттер для инкапсулированного поля цвета фигуры.
 * Реализация данного метода прописана в файле shape.cpp - это означает,
 * что данный метод напрямую можно вызывать от ссылки асбтрактного класса.
 * @return возвращает цвет данной фигуры.
 */
Colors Shape::GetColor() {
    return color;
}

/**
 * Сеттер цвета данной фигуры.
 * Данный метод также принадлежит обьекту асбтрактного класса.
 * Он позволяет установить цвет фигуры.
 * @param color - цвет, который нужно установить.
 */
void Shape::SetColor(Colors newColor) {
    color = newColor;
}

/**
 * Статический ввод фигуры.
 * На основе полученных данных из файла программа генерирует одну из альтернатив.
 * @param inputFile - файл, из которого происходит чтение.
 * @return Возвращает обобщенный экземпляр фигуры.
 */
Shape *Shape::StaticIn(FILE *inputFile) {
    int k;
    bool checkType = ProgramException::GetInt(inputFile, &k);

    int colorType;
    bool checkColor = ProgramException::GetInt(inputFile, &colorType);
    if (!checkType || !checkColor)
        return nullptr;
    Shape *sp = nullptr;
    switch (k) {
        case 1:
            sp = new Rectangle;
            break;
        case 2:
            sp = new Triangle;
            break;
        case 3:
            sp = new Circle;
            break;
        default:
            "Incorrect random value";
    }
    int *x = new int;
    sp->SetColor(Colors(colorType));
    sp->In(inputFile);

    delete x;
    return sp;
}

/**
 * Статический метод генерации рандомной алтернативы с рандомными параметрами.
 * Гарантируется, что какая бы алтернатива не была сгенерирована,
 * ее параметры будут отвечать математическим правилам.
 * @return Возвращает сгенерированный обобщенный экземпляр фигуры.
 */
Shape *Shape::StaticInRnd() {
    auto k = Random::next(1, 4);
    Shape *sp = nullptr;
    switch (k) {
        case 1:
            sp = new Rectangle;
            break;
        case 2:
            sp = new Triangle;
            break;
        case 3:
            sp = new Circle;
            break;
        default:
            "Incorrect random value";
    }
    sp->SetColor(Colors(Random::next(0, 7)));
    sp->InRnd();
    return sp;
}

File: ДЗ3 Хан Роман Олегович БПИ207/Program/shape.py
from enum import Enum


# Перечислимый тип цветов фигуры.
class Colors(Enum):
    TRANSPARENT = 0
    RED = 1
    ORANGE = 2
    YELLOW = 3
    GREEN = 4
    LIGHTBLUE = 5
    BLUE = 6
    PURPLE = 7


# Абстрактный класс фигуры.
class Shape:
    # Базовый цвет - прозрачный.
    # Здесь же и обеспечивается наличие общего свойства у всех алтернатив.
    def __init__(self):
        self.color = Colors.TRANSPARENT

    # Устанавливает цвет фигуры.
    def SetColor(self, i):
        self.color = i

    # Позволяет получить строкове представление цвета фигуры.
    def GetColorString(self) -> str:
        if self.color == 1:
            return "RED"
        elif self.color == 2:
            return "ORANGE"
        elif self.color == 3:
            return "YELLOW"
        elif self.color == 4:
            return "GREEN"
        elif self.color == 5:
            return "LIGHTBLUE"
        elif self.color == 6:
            return "BLUE"
        elif self.color == 7:
            return "PURPLE"
        else:
            self.color = 0
            return "TRANSPARENT"

    # Создание фигуры по данным из общего массива.
    def ReadStrArray(self, strArray, i):
        pass

    # Вывод данных о фигуре в консоль.
    def Print(self):
        pass

    # Вывод данных о фигуре в файл.
    def Write(self, ostream):
        pass

    # Вычисление периметра фигуры.
    def Perimeter(self):
        pass
